Services View: Added OpenShift Imagestreams

dtaylor113 opened the pull request:
Major enhancement of the Services View component which now displays Catalog Services alongside of OpenShift Imagestreams.   

I still need to reconcile [catalog categories](https://github.com/openshift/origin-web-catalog/blob/master/src/constants.ts#L12) with [consoles categories](https://github.com/openshift/origin-web-console/blob/master/app/scripts/constants.js#L301).  Need to deep-dive with someone and determine what else needs to be added to the Catalog's categories.

@jwforres , @spadgett, @jeff-phillips-18 

Fixes #24

jwforres commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -1,156 +1,173 @@
+    };
+
+    public closeOrderingPanel = () => {
+        this.ctrl.orderingPanelvisible = false;
```
nit but should be orderingPanelVisible with a capital V

jwforres replied:
Why do you need to track if the ordering panel is visible?

dtaylor113 replied:
This variable is used with the ng-if to hide/show the ordering panel.

jwforres replied:
Hmm I assumed the overlay was being done as a modal (like uibModal), I twill take a look at the overlay code and get back to you on this one.

jwforres commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -1,156 +1,173 @@
+    }
+
+    private getRowOfSubCategory(subCategory: string) {
+        for (let row = 0; row < this.ctrl.subCategories.length; row += 1) {
```
i know you said this code is changing, but a general comment is that we use _.each instead of for loops, the code is cleaner to read without all the array indexing

dtaylor113 replied:
Yes, I'm 99% sure this is going to go away, but point taken about always using _.each ;-)

jwforres commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -1,156 +1,173 @@
+    private normalizeData(type: string, items: any) {
+        let retSvcs = [];
+        let objClass: any;
+        _.each(items, (item: any, key: string) => {
```
if we aren't going to use an argument we generally don't define it, so in the case of an _.each we typically leave off key

dtaylor113 replied:
Fixed

jwforres commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -1,156 +1,173 @@
+                objClass = new OpenshiftImageItem(item);
+            }
+            // each item/card needs a call back to 'this' class
+            _.assign(objClass, {ctrl: this});
```
shouldn't give individual model objects access to the controller, why do the items need this?

dtaylor113 replied:
This is somewhat of a hack.  pf-card-view has an anonymous callback when a card is clicked, unfortunately the callback function cannot access the functions/variables of the ServiceViewController class, thus we stamp the 'this' in there.  
At one point I had this working without using the click callback, might be worth re-implementing that way?

jwforres replied:
I see, I'd rather pass references to functions than the entire controller.  How many functions is it needing access to?

jeff-phillips-18 replied:
If you define:

    public handleClick = (item: any, e: any) => {

wouldn't that give it access to 'this'?

dtaylor113 replied:
I had thought I tried that in the early stages, but I will try it again.  -thanks!

dtaylor113 replied:
Fixed

spadgett commented on app/pages/home/homePageController.ts:
```diff
@@ -30,17 +30,22 @@ export class HomePageController {
+        this.ctrl.catalogServices = resources.by("metadata.name");
       });
     } catch (e) {
       console.log("Error Loading serviceclasses from Service Catalog.k8s.io: " + e);
```
+1, I had opened #24 for this as well

dtaylor113 replied:
Fixed

spadgett commented on src/components/services-view/service-view.classes.ts:
```diff
@@ -0,0 +1,181 @@
+
+    private getName() {
+        if (this.data.metadata.annotations) {
+            return this.data.metadata.annotations['openshift.io/display-name'];
```
I noticed @jeff-phillips-18 added this filter in https://github.com/openshift/origin-web-common/pull/8

dtaylor113 replied:
Waiting on openshift/origin-web-common#8

dtaylor113 replied:
fixed

dtaylor113 commented on app/pages/home/homePageController.ts:
```diff
@@ -30,17 +30,22 @@ export class HomePageController {
+      this.ctrl.catalogServices = {};
     }
+
+    try {
```
Hi, yes, getting a lot of runtime errors when/if 'addtionalServers' isn't setup in config.js, or other settings in the pipeline are not configured correctly (etc/hosts).

jwforres replied:
Gotcha, that is a temporary issue while we are prototyping, for the real MVP (and beyond) we wont be using additionalServers as everything will be proxied behind the main API server hostname, can you add TODO comments to take these try/catches out?

spadgett replied:
@jwforres Should we change DataService so that passing an unrecognized resource doesn't cause an exception?

dtaylor113 replied:
Added the TODO comments

dtaylor113 commented on src/components/services-view/service-view.classes.ts:
```diff
@@ -0,0 +1,181 @@
+        this.description = this.getDescription();
+        this.longDescription = this.getLongDescription();
+        this.subCategory = this.getSubCategory();
+        this.category = this.getCategory();
```
Ugh, that would require a change to a lot of code in the service view.  How would an item/object be configured to be in multiple categories?  Would it have multiple 'tags' which equal multiple sub-categories?  I was afraid I didn't fully understand the categorization process yet :-(

jwforres replied:
yeah they can have many tags, and you might have tags that put something into more then one category.  Our existing catalog supports this.  As an example, I might have a template that is an example with a database and a php app, i might tag that both with "database" and "php" so that it shows up in both places.  The idea is to catch people from multiple points of entry since they might look for things in different ways.

jwforres replied:
also it isnt just multiple subcategories, it could be underneath multiple top level categories as well

dtaylor113 replied:
Ok, understood, this is going to take a little longer to address

dtaylor113 replied:
Hi, fixed ServiceClasses to be under multiple categories and sub-categories.  Waiting on Jeff's PR in order to use relevant filters for the builder images.

dtaylor113 commented on src/components/order-service/order-service.controller.ts:
```diff
@@ -12,10 +12,12 @@ export class OrderServiceController implements angular.IController {
-    this.ctrl.serviceName = this.ctrl.service.displayName ? this.ctrl.service.displayName : this.ctrl.service.metadata.name;
-    this.ctrl.versions = this.ctrl.service.versions;
-    this.ctrl.selectedVersion = this.ctrl.service.versions[0];
+    this.ctrl.serviceIcon = this.ctrl.service.icon ? this.ctrl.service.icon : 'pf pficon-service';
```
Is imageUrl under 'metadata' or 'osbMetadata'?   I can't find 'osbMetadata' definition in https://github.com/kubernetes-incubator/service-catalog/blob/master/pkg/apis/servicecatalog/v1alpha1/types.go#L133.  -thanks!

jwforres replied:
osbMetadata, its a raw extension blob which means that data is just getting passed back from the underlying info the broker sent to the service catalog API. You wont see the fields called out in the svc cat schema because of that.  There are no guarantees that any of that metadata will be there, its all just done by convention.

dtaylor113 replied:
Added 'imageUrl'

jwforres replied:
I see where this was fixed in the models and for the tiles, but in the order panel it looks like its still just using the class-based approach and not handling both ways.

dtaylor113 commented on app/mockServices/mockData/services.ts:
```diff
@@ -1,16 +1,15 @@
-    displayName: 'WildFly',
-    imageURL: 'font-icon icon-wildfly',
+    'spec': {
+      'tags': ['java']
```
Fixed

dtaylor113 commented on app/mockServices/mockData/services.ts:
```diff
@@ -1,16 +1,15 @@
+    'spec': {
+      'tags': ['java']
+    },
+    displayName: 'Test ServiceClass Java',
```
Fixed

dtaylor113 commented on app/pages/home/home-page.html:
```diff
@@ -6,7 +6,7 @@
   </landingheader>
   <landingbody>
-    <services-view services="$ctrl.services" categories="$ctrl.categories"></services-view>
+    <services-view catalog-services="$ctrl.catalogServices" openshift-image-streams="$ctrl.openshiftImageStreams"></services-view>
```
Fixed

dtaylor113 commented on app/pages/home/homePageController.ts:
```diff
@@ -14,7 +14,7 @@ export class HomePageController {
     this.ctrl.loading = true;
     this.ctrl.applications = [];
-    this.ctrl.services = [];
+    this.ctrl.catalogServices = [];
```
Fixed

dtaylor113 commented on src/components/services-view/service-view.classes.ts:
```diff
@@ -0,0 +1,181 @@
+
+    private getIcon() {
+        if (this.annotations) {
+            return 'font-icon ' + this.annotations.iconClass;
```
Waiting on openshift/origin-web-common#8

dtaylor113 replied:
Waiting on openshift/origin-web-common#8

dtaylor113 replied:
fixed

dtaylor113 commented:
@jeff-phillips-18 , do you have any idea why Travis is failing with:
    modified:   dist/vendor-bundle.js
    The command "hack/verify-dist.sh" exited with 1.
I noticed #21 failed with the same error.  I've run 'hack/install-deps' and 'verify-dist' locally and they run fine.

dtaylor113 commented:
Hi @jwforres, I believe all PR issues have been addressed

@spadgett, just a heads up that I changed webpack-dev.js's 'dev-tool' back to 'eval-source-maps' because 'inline-source-maps' (for some reason we are still investigating) caused webpack to pause for a really long time at '92% optimization chunks'.

jwforres commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -1,156 +1,190 @@
+            this.ctrl.filteredItems = this.ctrl.allItems;
         } else {
-          return  service.metadata.category === category && service.metadata.subCategory === subCategory;
+            this.ctrl.filteredItems = this.$filter('filter')(this.ctrl.allItems, (item: any) => {
```
we prefer to use lodash's _.filter from the JS instead of $filter('filter')

dtaylor113 replied:
Fixed

jwforres commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -1,156 +1,190 @@
+        this.updateState();
+    }
+
+    private UpdateImageStreams() {
```
use consistent casing, you have update above and Update on this one

dtaylor113 replied:
Fixed

jwforres commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -1,156 +1,190 @@
+    }
+
+    private hasCategory(item: any, category: string) {
+        return _.some(item.catsBySubCats, (itemCategory: string) => {
```
```
return _.includes(item.catsBySubCats, category);
```

dtaylor113 replied:
Fixed

jwforres commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -1,156 +1,190 @@
+    }
+
+    private hasSubCategory(item: any, subCategory: string) {
+        return _.some(Object.keys(item.catsBySubCats), (itemSubCategory: string) => {
```
same as above, use _.includes

dtaylor113 commented on app/mockServices/mockData/services.ts:
```diff
@@ -1,16 +1,16 @@
 export const servicesData = {
-  "wildfly": {
+  "test-serviceclass-java": {
```
Arrgg...my apologies!  I meant to remove 'service-view.classes.ts'!  It has been replaced with 'src/services/catalog.service.ts'

dtaylor113 commented on src/components/services-view/service-view.classes.ts:
```diff
@@ -0,0 +1,193 @@
+        this.longDescription = this.getLongDescription();
+        this.catsBySubCats = this.getCategoriesBySubCategories();
+        this.$filter = $filter;
+        console.log("construct $filter = " + JSON.stringify($filter, null, ' '));
```
Arrgg...my apologies!  I meant to remove 'service-view.classes.ts'!  It has been replaced with 'src/services/catalog.service.ts'

jwforres commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,198 @@
+    return new ImageItem(resource, this);
+  }
+
+  public getCategoryFromSubCategory(subCategory: string) {
```
I had several comments about these methods against the other file, i think they all still apply here if you can just look back at those in the history.

dtaylor113 replied:
Hi, these methods were removed and combined into 'getCategoriesBySubCategories'.

jwforres commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,198 @@
+      return this.resource.osbMetadata.displayName ? this.resource.osbMetadata.displayName : this.resource.metadata.name;
+    } else {
+      return this.resource.metadata.name;
+    }
```
this whole if/else can be simplified down to

```
return _.get(this.resource, 'osbMetadata.displayName', this.resource.metadata.name);
```

dtaylor113 replied:
Fixed

jwforres commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,198 @@
+  private getDescription() {
+    if (this.resource.osbMetadata) {
+      return this.resource.osbMetadata.description;
+    }
```
remove the if:
```
return _.get(this.resource, 'osbMetadata.description');
```

dtaylor113 replied:
Fixed

jwforres commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,198 @@
+
+  private getLongDescription() {
+    if (this.resource.osbMetadata) {
+      return this.resource.osbMetadata.longDescription;
```
same thing, _.get

dtaylor113 replied:
Fixed

jwforres commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,198 @@
+  private getCategoriesBySubCategories() {
+    let catsBySubCats = {};
+
+    if (this.resource.osbTags) {
```
the if isn't necessary, _.each already handles existence checks on what you pass to it so if this.resource.osbTags is null or undefined it won't care it just treats it like there are no items

dtaylor113 replied:
Fixed

dtaylor113 replied:
Fixed

jwforres commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,198 @@
+  constructor (serviceClass: any, catalogSrv: CatalogService) {
+    this.resource = serviceClass;
+    this.catalogSrv = catalogSrv;
+    this.imageUrl = this.getImage();
```
a service class might have either an imageUrl or an iconClass, our own brokers we write specifically for openshift will use iconClass, although there was some discussion about prefixing icon class with something ('something.openshift.io/iconClass'), @bparees did you guys decide what this will be?

bparees replied:
i'm not sure we picked for certain, but i think it'll be `template.openshift.io/iconClass`.  @jim-minter can confirm/deny my recollection.

bparees replied:
> I wonder if that's the right prefix if we want other brokers we write to
use the same annotation.  Maybe we should do something like
console.openshift.io since it's really a convention for the console?

makes sense.  i could also see broker.openshift.io/iconClass, though it is a bit redundant since anything providing that value is, by definition, a broker.

jim-minter replied:
the commit had `template.openshift.io/iconClass` but I agree it makes sense to change it.  Let me know what you choose (`console.openshift.io/iconClass` sounds pretty sensible to me) and I can do a fix-up on the template service broker side if you want me to.

bparees replied:
console.openshift.io/iconClass works for me. let's do it.

jwforres commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,198 @@
+  }
+
+  private getIcon() {
+    return 'font-icon ' + this.catalogSrv.$filter('imageStreamTagIconClass')(this.resource);
```
repeating this comment that i put on the old file:
so i think i said before tagName was optional but I just realized that when you don't include it then it is assuming the tag is "latest" which we shouldn't really make the assumption that a latest tag exists :-\

so i think you need to actually grab the name of a valid tag off this imagestream, i think you can just do what you do above in annotations and just use the first tag you find in the list to get the name of a valid tag

dtaylor113 replied:
Fixed

jwforres commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,198 @@
+  }
+
+  private getName() {
+    return this.catalogSrv.$filter('annotation')(this.resource, "displayName");
```
should fallback to this.resource.metadata.name when displayName doesn't exist

spadgett replied:
@dtaylor113 If you use the `displayName` filter instead of `annotation`, it will do that for you.

```js
return this.catalogSrv.$filter('displayName')(this.resource);
```

dtaylor113 replied:
Fixed in next commit.

jwforres commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,198 @@
+    let catsBySubCats = {};
+    let tags =  this.catalogSrv.$filter('imageStreamTagTags')(this.resource);
+
+    this.isBuilder = _.some(tags, (tag) => {
```
repeating from the other file:
its weird that a "getter" is also modifying the object. I think instead I would make "tags" a private instance variable and set it to an array that you generate once by splitting the tags string. Then you can have a simple line after that in the constructor for this.isBuilder = _.includes(this.tags, "builder") and then have this method use the already split tags to figure out the categories.

jwforres commented:
I wonder if that's the right prefix if we want other brokers we write to
use the same annotation.  Maybe we should do something like
console.openshift.io since it's really a convention for the console?

On Mar 22, 2017 8:05 PM, "Ben Parees" <notifications@github.com> wrote:

*@bparees* commented on this pull request.
------------------------------

In src/services/catalog.service.ts
<https://github.com/openshift/origin-web-catalog/pull/19#discussion_r107562517>
:

> +}
+
+export class ServiceItem implements IServiceItem {
+  public iconClass: string;
+  public imageUrl: string;
+  public name: string;
+  public catsBySubCats: any;
+  public description: string;
+  public longDescription: string;
+  public resource: any;
+  private catalogSrv: CatalogService;
+
+  constructor (serviceClass: any, catalogSrv: CatalogService) {
+    this.resource = serviceClass;
+    this.catalogSrv = catalogSrv;
+    this.imageUrl = this.getImage();

i'm not sure we picked for certain, but i think it'll be
template.openshift.io/iconClass. @jim-minter <https://github.com/jim-minter>
can confirm/deny my recollection.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/openshift/origin-web-catalog/pull/19#discussion_r107562517>,
or mute the thread
<https://github.com/notifications/unsubscribe-auth/ABZk7QezEHjejfUgmQDlLpP5UMbGeYWvks5robddgaJpZM4Mf63b>
.

dtaylor113 commented on src/components/services-view/service-view.classes.ts:
```diff
@@ -0,0 +1,193 @@
+            });
+            let lastTag = tags.pop();
+            let subCat = getSubCategoryFromAlias(lastTag);
+            let cat = getCategoryFromSubCategory(subCat);
```
Hi, fixed this in ''src/services/catalog.service.ts'

dtaylor113 commented on src/components/services-view/service-view.classes.ts:
```diff
@@ -0,0 +1,193 @@
+    let retCategory: string = 'other';
+
+    let breakLoops: boolean = false;
+    _.each(categories, (category) => {
```
Fixed this in ''src/services/catalog.service.ts', although I got a runtime error when I tried to use _.findWhere, but it should exist, we are using lodash 3.10.1?  I used _.findIndex instead.

dtaylor113 commented on src/components/services-view/service-view.classes.ts:
```diff
@@ -0,0 +1,193 @@
+    let retSubCategory: string = subCategory;
+
+    let breakLoops: boolean = false;
+    _.each(categories, (category) => {
```
Fixed this in ''src/services/catalog.service.ts'

dtaylor113 commented on src/components/services-view/service-view.classes.ts:
```diff
@@ -0,0 +1,193 @@
+            return false;
+        }
+        _.each(category.subCategories, (cat_subCategory) => {
+            if (cat_subCategory.categoryAliases) {
```
Fixed this in ''src/services/catalog.service.ts'

dtaylor113 commented on src/components/services-view/service-view.classes.ts:
```diff
@@ -0,0 +1,193 @@
+            if (cat_subCategory.categoryAliases) {
+                let result = _.findIndex(cat_subCategory.categoryAliases, (alias) => {
+                    return alias === subCategory;
+                });
```
Fixed this in ''src/services/catalog.service.ts'

jwforres commented:
I think console.openshift.io is the right prefix, and then the annotation
makes sense even outside the service catalog use case.

On Mar 23, 2017 5:10 AM, "Jim Minter" <notifications@github.com> wrote:

> *@jim-minter* commented on this pull request.
> ------------------------------
>
> In src/services/catalog.service.ts
> <https://github.com/openshift/origin-web-catalog/pull/19#discussion_r107617406>
> :
>
> > +}
> +
> +export class ServiceItem implements IServiceItem {
> +  public iconClass: string;
> +  public imageUrl: string;
> +  public name: string;
> +  public catsBySubCats: any;
> +  public description: string;
> +  public longDescription: string;
> +  public resource: any;
> +  private catalogSrv: CatalogService;
> +
> +  constructor (serviceClass: any, catalogSrv: CatalogService) {
> +    this.resource = serviceClass;
> +    this.catalogSrv = catalogSrv;
> +    this.imageUrl = this.getImage();
>
> the commit had template.openshift.io/iconClass but I agree it makes sense
> to change it. Let me know what you choose (console.openshift.io/iconClass
> sounds pretty sensible to me) and I can do a fix-up on the template service
> broker side if you want me to.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/origin-web-catalog/pull/19#discussion_r107617406>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZk7ZFZPOoWdThHD2pmIbabY1ouKHfKks5rojbkgaJpZM4Mf63b>
> .
>

dtaylor113 commented on app/mockServices/mockData/services.ts:
```diff
@@ -1,16 +1,16 @@
-      subCategory: 'java',
+      "name": "test-serviceclass-java-nodejs",
       "uid": "1",
+      "imageUrl": "http://www.rightstar.com/wp-content/uploads/2014/03/Java.png"
```
Added issue #28 , I will remove the urls

spadgett commented on app/pages/home/homePageController.ts:
```diff
@@ -25,22 +28,30 @@ export class HomePageController {
 
   public update() {
+
+    // TODO: Remove these try/catches once we are not relying on 'additionalServers' in config.js
```
@jwforres I think unrecognized resource should be a promise failure instead of runtime error. Then if the service catalog is down or not installed, it can use our existing error handling.

jwforres replied:
@spadgett opened an issue in common to track it, watch is a problem still https://github.com/openshift/origin-web-common/issues/17

dtaylor113 replied:
Fixed

spadgett commented on app/pages/services/servicesPageController.ts:
```diff
@@ -25,21 +28,28 @@ export class ServicesPageController {
-        this.ctrl.categories = this.constants.SERVICE_CATALOG_CATEGORIES;
-        this.ctrl.loading = false;
+        this.ctrl.serviceClasses = resources.by("metadata.name");
+      });
```
@dtaylor113 Your calls to `dataService.list` should use the promise returned rather than passing in the callback as a parameter.

jwforres replied:
yeah agreed switch this to use the promise like you did in homePageController

dtaylor113 replied:
Fixed

spadgett commented on src/components/services-view/services-view.component.ts:
```diff
@@ -2,8 +2,8 @@ import {ServicesViewController} from './services-view.controller';
-    services: '<',
-    categories: '<'
+    serviceClasses: '<',
+    imageStreams: "<"
```
Nit: let's consistently use either single quote or double quote styles in the same declaration

dtaylor113 replied:
Fixed

spadgett commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -2,155 +2,188 @@ import * as angular from 'angular';
+        this.ctrl.expandSubCatRow = this.getRowOfSubCategory(subCategory);
+
+        this.ctrl.currentFilter = category;
+        this.ctrl.currentSubFilter = (subCategory !== undefined) ? subCategory : 'all';
```
Any reason not to use...?

```js
this.ctrl.currentSubFilter = subCategory || 'all';
```

dtaylor113 replied:
Fixed

spadgett commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -2,155 +2,188 @@ import * as angular from 'angular';
-          return service.metadata.subCategory === subCategory;
+
+    public $onChanges(onChangesObj: angular.IOnChangesObject) {
+        if (onChangesObj.serviceClasses && !onChangesObj.serviceClasses.isFirstChange()) {
```
If the service classes or image streams are loaded before the directive is used, will that mean that the directive is never initialized?

dtaylor113 replied:
Hi, please see UpdateState() which tests for empty or if the particular resources has been 'loaded'.  This is to handle when the directive is initialized with empty serviceClasses/imageStreams and is waiting on the DataService to return asynchronously.

spadgett replied:
It looks like `this.ctrl.loading` will never be set to true in `updateState()` if image streams or service classes are empty?

dtaylor113 replied:
I believe the only way this.ctrl.loading will never be set to false (what we need to stop the spinners) if images streams or service classes are initially set to empty, and never updated again asynchronously.  After DataService.list sets those collections asynchronously (to empty obj. or otherwise) the flags (serviceClassesLoaded & imageStreamsLoaded) will be set to true and this.ctrl.loading will be set to false, thus stopping the spinners.

spadgett commented on src/components/services-view/services-view.controller.ts:
```diff
@@ -2,155 +2,188 @@ import * as angular from 'angular';
+        }
+    }
+
+    public filterByCategory(category: string, subCategory: string, updateSubCategories: boolean) {
```
The approach we took with the previous catalog is to create maps of category and subcategories with the items once on initialization. Then we don't have to filter the lists every time the user clicks a tile.

dtaylor113 replied:
I'll review that code and follow up with another PR.

spadgett commented on src/components/services-view/services-view.html:
```diff
@@ -35,10 +35,11 @@
           <div class="card-container">
-            <div class="card-icon {{item.imageURL ? item.imageURL : 'pf pficon-service'}}"></div>
-            <div class="card-name">{{item.displayName ? item.displayName : item.metadata.name}}</div>
+            <div ng-if="!item.imageUrl" class="card-icon {{item.iconClass ? item.iconClass : 'pf pficon-service'}}"></div>
```
Class `pf`? Should that be `pficon`? Not sure if `pficon-service` is really the default we want.

dtaylor113 replied:
This has been removed in my local branch, not committed yet.

spadgett commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,166 @@
+    _.each(tags, (tag) => {
+      _.each(this.categories, (category) => {
+        let subCat: any = _.find(category.subCategories, (subCategory: any) => {
+              if (subCategory.id === tag) {
```
indentation

spadgett commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,166 @@
+              } else if (subCategory.categoryAliases) {
+                return _.includes(subCategory.categoryAliases, tag);
+              }
+            }
```
I prefer the shorter

```js
  return subCategory.id === tag || _.includes(subCategory.categoryAliases, tag);
```

Lodash should handle undefined category aliases

dtaylor113 replied:
Fixed

spadgett commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,166 @@
+              }
+            }
+        );
+        if (!_.isUndefined(subCat)) {
```
Any reason not to use?

```js
if (subCat)
```

dtaylor113 replied:
Fixed

jwforres commented on app/pages/home/homePageController.ts:
```diff
@@ -25,22 +28,26 @@ export class HomePageController {
-      this.ctrl.loading = false;
-    }
+
+    // TODO: Remove these try/catches once we are not relying on 'additionalServers' in config.js
```
since you switched to the promises can you remove this comment, since handling the promise rejection is a good thing to keep in case we ever get errors from the API or the server goes down

dtaylor113 replied:
Fixed

dtaylor113 commented on app/pages/services/servicesPageController.ts:
```diff
@@ -25,21 +28,28 @@ export class ServicesPageController {
+        this.ctrl.serviceClasses = resources.by("metadata.name");
+      });
+    } catch (e) {
+      this.logger.log("Error Loading serviceclasses from Service Catalog.k8s.io: " + e);
```
Fixed

dtaylor113 commented on src/components/services-view/service-view.classes.ts:
```diff
@@ -0,0 +1,193 @@
+
+    private getIcon() {
+        console.log("$filter = " + JSON.stringify(this.$filter, null, ' '));
+        let icon = this.$filter('imageStreamTagIconClass')(this.data, "icon");
```
Fixed

dtaylor113 commented on src/components/services-view/service-view.classes.ts:
```diff
@@ -0,0 +1,193 @@
+        let catsBySubCats = {};
+        if (this.annotations) {
+            let tags = this.annotations.tags.split(/\s*,\s*/);
+            this.isBuilder = _.some(tags, (tag) => {
```
Fixed

dtaylor113 commented on src/components/services-view/service-view.classes.ts:
```diff
@@ -0,0 +1,193 @@
+            this.isBuilder = _.some(tags, (tag) => {
+                return tag === 'builder';
+            });
+            let lastTag = tags.pop();
```
Fixed

jwforres commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,152 @@
+    let catsBySubCats = {};
+    _.each(tags, (tag) => {
+      _.each(this.categories, (category) => {
+        let subCat: any = _.find(category.subCategories, (subCategory: any) => {
```
you know what, just realized if this matches multiple subcategories only the first one will actually get matched, i dont think we can break out early at all here

dtaylor113 replied:
I don't follow.  A single tag like 'nodejs' is going to match a single sub-category (or alias).  A service item may have multiple tags, which may match to multiple categories and sub-categories.   We don't have a mapping of matching a single tag to multiple sub-categories?

dtaylor113 replied:
This may come back to my initial comment: "I still need to reconcile [catalog categories](https://github.com/openshift/origin-web-catalog/blob/master/src/constants.ts#L12) with [consoles categories](https://github.com/openshift/origin-web-console/blob/master/app/scripts/constants.js#L301).  Need to deep-dive with someone and determine what else needs to be added to the Catalog's categories."

I see in  [consoles categories](https://github.com/openshift/origin-web-console/blob/master/app/scripts/constants.js#L301) there is something called 'subcategories', which really makes that structure 3 levels deep:  Categories -> Items -> subcategories.  I am currently using a categories constant which is two levels deep.

dtaylor113 replied:
Either way, it would be great if we could handle this in a follow up PR.

jwforres commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  private getIcon() {
+    return 'fa fa-cubes';
```
so based on the discussion with API folks, i think this becomes
```
return _.get(this.resource, ['osbMetadata', 'console.openshift.io/iconClass'], 'fa fa-cubes');
```
And we can debate later if there is a better default icon for service class items vs what we use for builder images.

dtaylor113 replied:
Ok, thanks, fixed

jwforres commented on src/services/catalog.service.ts:
```diff
@@ -0,0 +1,152 @@
+  }
+
+  private getImage() {
+    return _.get(this.resource, 'metadata.imageUrl', '');
```
osbMetadata not metadata

dtaylor113 replied:
Fixed

dtaylor113 commented:
Hi @jwforres ,
So for service class icons, is 'console.openshift.io/IconClass' under 'osbMetadata'?
Thanks

dtaylor113 commented:
Hi @jwforres & @spadgett,  I believe I've addressed all issues (except the last one @jwforres raised, but hoping to push that to a follow up PR).
Thanks,
- Dave

jwforres commented:
Oh an please squash the commits :)

dtaylor113 commented:
Created issue #29 so we can follow up on the subcategory matching logic
